### PR TITLE
Node-local batch normalization

### PR DESCRIPTION
**Pull request by ndryden**
Adds support for node-local batch normalization. Instead of `global_stats`, batchnorm now has a suite of aggregation options: `local` (the default, as before), `nodelocal`, and `global`.

### CODE REVIEW

File: include/lbann/layers/regularizers/batch_normalization.hpp [hunk 1]
```diff
@@ -32,6 +32,15 @@
 
 namespace lbann {
 
+enum class batch_normalization_stats_aggregation {
+  /** Statistics are aggregated only within a single rank. */
+  local,
+  /** Statistics are aggregated among every rank in a single node. */
+  nodelocal,
```
Comment (timmoon10) [inline include/lbann/layers/regularizers/batch_normalization.hpp]: `node_local` matches our coding style better.

File: src/layers/regularizers/batch_normalization.cpp [hunk 2]
```diff
@@ -73,10 +73,20 @@ void batch_normalization_layer<data_layout::DATA_PARALLEL, El::Device::CPU>::fp_
       local_var(channel, 0) = sqsum;
     }
     El::Int num_per_sum;
-    if (m_use_global_stats) {
+    if (m_stats_aggregation == batch_normalization_stats_aggregation::global) {
```
Comment (timmoon10) [inline src/layers/regularizers/batch_normalization.cpp]: I think a switch-case statement would be easier to read.

File: src/layers/regularizers/batch_normalization.cpp [hunk 3]
```diff
@@ -208,13 +218,20 @@ void batch_normalization_layer<data_layout::DATA_PARALLEL, El::Device::CPU>::bp_
 
   // Accumulate gradients
   if (is_training) {
-    if (m_use_global_stats) {
+    if (m_stats_aggregation == batch_normalization_stats_aggregation::global) {
```
Comment (timmoon10) [inline src/layers/regularizers/batch_normalization.cpp]: Switch-case statement?

File: src/layers/regularizers/batch_normalization.cpp [hunk 4]
```diff
@@ -232,9 +249,14 @@ void batch_normalization_layer<data_layout::DATA_PARALLEL, El::Device::CPU>::bp_
   }
 
   // Compute error signal
-  const auto& num_per_sum = (m_use_global_stats ?
-                             width * channel_size :
-                             local_width * channel_size);
+  El::Int num_per_sum;
+  if (m_stats_aggregation == batch_normalization_stats_aggregation::global) {
```
Comment (timmoon10) [inline src/layers/regularizers/batch_normalization.cpp]: Switch-case statement?

File: src/layers/regularizers/batch_normalization.cu [hunk 5]
```diff
@@ -339,10 +339,20 @@ void batch_normalization_layer<data_layout::DATA_PARALLEL, El::Device::GPU>::fp_
           local_mean.Buffer(), local_var.Buffer());
     }
     El::Int num_per_sum;
-    if (m_use_global_stats) {
+    if (m_stats_aggregation == batch_normalization_stats_aggregation::global) {
```
Comment (timmoon10) [inline src/layers/regularizers/batch_normalization.cu]: Switch-case statement?

File: src/layers/regularizers/batch_normalization.cu [hunk 6]
```diff
@@ -447,13 +457,20 @@ void batch_normalization_layer<data_layout::DATA_PARALLEL, El::Device::GPU>::bp_
 
   // Accumulate gradients
   if (is_training) {
-    if (m_use_global_stats) {
+    if (m_stats_aggregation == batch_normalization_stats_aggregation::global) {
```
Comment (timmoon10) [inline src/layers/regularizers/batch_normalization.cu]: Switch-case statement?

File: src/layers/regularizers/batch_normalization.cu [hunk 7]
```diff
@@ -471,9 +488,14 @@ void batch_normalization_layer<data_layout::DATA_PARALLEL, El::Device::GPU>::bp_
   }
 
   // Compute error signal
-  const auto& num_per_sum = (m_use_global_stats ?
-                             width * channel_size :
-                             local_width * channel_size);
+  El::Int num_per_sum;
+  if (m_stats_aggregation == batch_normalization_stats_aggregation::global) {
```
Comment (timmoon10) [inline src/layers/regularizers/batch_normalization.cu]: Switch-case statement?

File: src/layers/regularizers/batch_normalization.cpp (hunk 3 above)
Comment (ndryden) [reply src/layers/regularizers/batch_normalization.cpp]: I don't think a switch statement when there are precisely two cases is worth while.

**Comment by ndryden** [comment]
All models using global batchnorm stats are updated.
